### PR TITLE
Comprehensive Picus determinism verification with blocking component + monolithic CI gates

### PR DESCRIPTION
**Pull request by adamkrellenstein**
## Summary
This PR introduces a production-oriented Picus formal-verification system for Kontor-Crypto and wires it into CI as blocking gates.

It verifies determinism / underconstraint safety in two complementary lanes:

1. `component_matrix` (modular, production-gadget subcircuits)
2. `monolithic_matrix` (broad full `PorCircuit` fixtures)

This supersedes the earlier monolithic-only flow with broader coverage, stronger negative controls, and fail-closed policy enforcement.

## Why this matters
The goal is to prevent underconstraints from shipping by continuously checking:
- expected-safe fixtures classify as `safe`
- expected-mutant fixtures classify as `unsafe`
- inconclusive/error outcomes fail CI

### 1. Formal architecture and tooling
- Added formal core module and fixture pipeline:
  - `src/formal/mod.rs`
  - `src/formal/components.rs`
  - `src/formal/roles.rs`
- Added dedicated CLIs:
  - `src/bin/picus_export.rs`
  - `src/bin/picus_verify.rs`

### 2. Component verification lane (blocking)
- Added component circuits built from production gadgets:
  - `src/circuit/formal_components.rs`
- Added contracts + composition checks:
  - `tools/picus/components/contracts.json`
- Added component fixture manifest + fixtures:
  - `tools/picus/components/manifest.json`
  - `tools/picus/components/fixtures/*.json`
- CI-fast profile includes safe + mutant controls across all 5 component families.

### 3. Monolithic broad lane (blocking)
- Added broad full-circuit manifest:
  - `tools/picus/manifest-broad.json`
- Broad profile uses full output scope (`output_prefix_len = 0`) in fixture policy.
- Includes monolithic mutant positive control:
  - `single-file-minimal-carry-forward-mutant` (expected `unsafe`)

### 4. Policy and regression guards
- Fixture-level `verification` policy supports:
  - `scope`
  - `output_prefix_len`
  - `solver_policy` (`primary`, `fallbacks`)
  - `timeout_secs`
  - `hard_timeout_grace_secs`
- `expected_num_constraints` is enforced during export (constraint-count drift guard).
- Strict `leafpath` scope is default fail-closed behavior.

### 5. CI workflow
- Added comprehensive workflow:
  - `.github/workflows/formal-picus.yml`
- Two blocking jobs:
  - `component_matrix`
  - `monolithic_matrix`
- Dockerized Picus integration and artifact upload included.

### 6. Reliability fixes in verifier (this branch)
- Fixed unsafe/safe classification ordering bug that could misclassify `unsafe` as `safe` due substring matching:
  - `src/bin/picus_verify.rs`
- Fixed hard-timeout process handling to drain stdout/stderr concurrently and avoid child pipe deadlock:
  - `src/bin/picus_verify.rs`
- Added regression tests for classifier edge cases.

### 7. Documentation
Updated and consolidated formal docs:
- `README.md`
- `docs/formal/README.md`
- `docs/formal/determinism.md`
- `docs/formal/components.md`
- `docs/formal/picus.md`
- `docs/formal/results.md`

## Determinism statement enforced by CI
For each blocking fixture, under configured scope and target outputs:
- expected-safe fixtures must classify `safe`
- expected-unsafe fixtures must classify `unsafe`
- inconclusive/error outcomes fail the lane

## Non-goals / limits
This PR does **not** prove:
- cryptographic assumptions (e.g. hash injectivity)
- full recursive Nova proof system soundness
- universal coverage beyond the provided fixture sets

## Validation performed
- `cargo fmt --all`
- `cargo check --all-targets`
- `cargo test --bin picus_verify` (added classifier regression tests)
- local Picus fixture runs via manifests and targeted checks

## Operational notes
- Requires Veridise Picus (`run-picus`), not PyPI `picus==0.0.5`.
- cvc5 should be CoCoA-enabled for `QF_FF`.
- `z3` is configured as fallback where specified by fixture policy.

---

> [!NOTE]
> **Medium Risk**
> Adds new formal-verification export/verify tooling plus CI workflows, and modifies core `PorCircuit` synthesis to optionally record witness aux indices; mistakes here could affect circuit behavior or formal gate reliability.
> 
> **Overview**
> Adds a **blocking Picus formal-verification pipeline** to check determinism/underconstraint safety across both *component sub-circuits* and *broad monolithic `PorCircuit` fixtures*, including a new scheduled/PR GitHub Actions workflow that runs a two-lane matrix and uploads verification artifacts.
> 
> Introduces new CLIs `picus_export` and `picus_verify` that load fixture manifests, export R1CS + optional preconditions, run Picus with per-fixture policy (scope/output targets/timeouts/solver fallbacks), and **fail closed** on unexpected `safe`/`unsafe` outcomes or inconclusive results; verifier reliability is improved with safer result classification, hard timeouts that kill process groups, and generated JSON/Markdown summaries.
> 
> Adds component-focused circuits (`formal_components`) and dev-only “lite” circuits behind `formal-dev`, plus contract validation (`src/formal/components.rs`) to enforce component interface role compatibility; also extends `PorCircuit` synthesis with optional witness tracing (`PorWitnessTrace`) to support `leafpath` preconditions.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 2911ae817cb4341245f8a3f671325f0192a298f8. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/circuit/synth.rs [hunk 1]
```diff
@@ -33,6 +50,101 @@ pub fn synthesize_por_circuit<F: PrimeField + PrimeFieldBits, CS: ConstraintSyst
...
+        trace.take(),
+        false,
+    )
+}
+
+/// Synthesize a monolithic positive-control mutant for formal verification.
+///
+/// The mutant intentionally emits an unconstrained output vector with the correct arity so
+/// Picus should quickly report `unsafe`.
+pub fn synthesize_por_circuit_carry_forward_mutant<
```
Comment (cryptograph) [inline src/circuit/synth.rs]: Looks like it is used only to test that picus returns unsafe when there is an unconstrained output. I don't like having even theoretical possibility of dropping constraints in prod circuit.

Let's remove `synthesize_por_circuit_carry_forward_mutant` and `drop_carry_forward_equalities` from the production circuit and create a simple control mutant with same arity as PorCircuit but unconstrained outputs in `formal_components.rs`

**Review by cryptograph — DISMISSED** [review]

**Comment by cryptograph** [comment]
why did you close https://github.com/KontorProtocol/Kontor-Crypto/pull/37 ?

**Comment by adamkrellenstein** [comment]
> why did you close #37 ?

Whoops! Good catch! just merged it

**Review by cryptograph — DISMISSED** [review]

**Review by cryptograph — APPROVED** [review]
